### PR TITLE
Support for virtIO sound

### PR DESCRIPTION
**Pull request by alexandermbrown**
This PR adds a virtIO sound device, along with an example Linux userlevel sound driver.

### CODE REVIEW

File: tools/linux/snd/asound_qemu_arm_virt.conf
Comment (Ivan-Velickovic) [inline tools/linux/snd/asound_qemu_arm_virt.conf:1]: If there is no specific configuration can you please add a comment in this file saying so just so this empty file doesn't cause confusion when we stumble upon it in the future.

File: tools/linux/uio_drivers/snd/queue.c [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+struct queue
+{
+    int head;
+    int len;
+    int capacity;
+    int item_size;
+    void *data;
+};
+
+static int min(int a, int b)
```
Comment (Ivan-Velickovic) [inline tools/linux/uio_drivers/snd/queue.c:15]: Doesn't seem to be used.

File: examples/virtio-snd/.clang-format
Comment (Ivan-Velickovic) [inline examples/virtio-snd/.clang-format:1]: I think this was accidentally committed?

File: examples/virtio-snd/snd_driver_vmm.c [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    assert(success);
+
+#if defined(BOARD_qemu_arm_virt)
+    register_passthrough_irq(37, 5);
+#elif defined(BOARD_odroidc4)
+    register_passthrough_irq(48, 5);
+    register_passthrough_irq(63, 6);
+    register_passthrough_irq(62, 7);
+    register_passthrough_irq(5, 8);
+#else
```
Comment (Ivan-Velickovic) [inline examples/virtio-snd/snd_driver_vmm.c:203]: These IRQ numbers should all have names, I shouldn't have to refer to the DTS to figure out what they are.

File: .gitmodules [hunk 3]
```diff
@@ -1,4 +1,4 @@
 [submodule "sddf"]
 	path = examples/virtio/sddf
 	url = https://github.com/au-ts/sddf
-	branch = main
+	branch = main
```
Comment (Ivan-Velickovic) [inline .gitmodules:4]: Remove this unrelated change.

**Review by Ivan-Velickovic — CHANGES_REQUESTED** [review]
I have written some minor things down to change but the major thing seems to be that this will need to rebase once Eric's block PR is merged. It's a bit hard to review with all of his changes in this PR as well (my fault, should've been quicker about merging things).

The `SOUND.md` documentation is good, but we will need to put the relevant bits in the right place. Some of it belongs in the README for the virtio example, some of it belongs in the manual.

I see that the `virtio-snd` example still exists and is not merged with the `virtio` example? Are you waiting for Eric's PR to go through first?

I think we discussed offline about how legally we can't include some of the example sound files due to copyright? We should fix that as well.

The design and implementation of the user-level side and virtIO device we can discuss together offline since it's easier.

File: examples/virtio-snd/.clang-format
Comment (alexandermbrown) [reply examples/virtio-snd/.clang-format:1]: Yeah this was from before we had the seL4-tools formatter. Are we planning on adding that to this repo's CI?

File: .gitmodules (hunk 3 above)
Comment (alexandermbrown) [reply .gitmodules:4]: Just checked, this file is identical to `main` so I don't think I actually modified anything

File: examples/virtio-snd/snd_driver_vmm.c (hunk 2 above)
Comment (alexandermbrown) [reply examples/virtio-snd/snd_driver_vmm.c]: I labelled them with comments in the most recent commit. Not sure if its worth putting # defines since the amount of interrupt types changes per device class.

File: tools/linux/uio_drivers/snd/queue.c (hunk 1 above)
Comment (alexandermbrown) [reply tools/linux/uio_drivers/snd/queue.c]: done

File: docs/SOUND.md
Comment (Ivan-Velickovic) [inline docs/SOUND.md:1]: Put this in `examples/virtio-snd` for now.

File: .gitignore [hunk 4]
```diff
@@ -4,3 +4,6 @@ zig-cache/
 *.pdf
 *.dtb
 rust/target/
+*overlayed.cpio.gz
```
Comment (Ivan-Velickovic) [inline .gitignore:7]: Do we need this in the gitignore? If it's a build artefact I would think it would be in the build directory right?

File: tools/linux/uio_drivers/snd/main.c [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+#include <uio/sound.h>
+#include <sddf/sound/queue.h>
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+#define PAGE_SIZE_4K 0x1000
+#define QUEUE_BYTES 0x200000
```
Comment (Ivan-Velickovic) [inline tools/linux/uio_drivers/snd/main.c:13]: This is dependent on the system setup right? If so can you make a comment here.

File: src/util/buffer.h
Comment (Ivan-Velickovic) [inline src/util/buffer.h:1]: I can't find anywhere where this file is used.

File: src/virtio/virtio.h [hunk 6]
```diff
@@ -1,30 +1,5 @@
...
-#include <stdbool.h>
-#include <stdint.h>
-#include <stddef.h>
-#include "virtio/mmio.h"
-
-/*
- * All terminology used and functionality of the virtIO device implementation
- * adheres with the following specification:
- * Virtual I/O Device (VIRTIO) Version 1.2
- */
```
Comment (Ivan-Velickovic) [inline src/virtio/virtio.h:17]: Why was this removed?

File: src/virtio/sound.c [hunk 7]
```diff
@@ -0,0 +1,851 @@
...
+    return 1;
+}
+
+static int virtio_snd_mmio_set_device_config(struct virtio_device *dev, uint32_t offset, uint32_t val)
+{
+    LOG_SOUND_ERR("Not allowed to change sound config.");
+    return 0;
+}
+
+// static const char *queue_name[] = { "controlq", "eventq", "txq", "rxq", NULL };
```
Comment (Ivan-Velickovic) [inline src/virtio/sound.c:100]: Is this still necessary?

File: src/virtio/virtio.h (hunk 6 above)
Comment (alexandermbrown) [reply src/virtio/virtio.h:17]: I can add it back haha

File: examples/virtio-snd/snd_driver_vmm.c [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+// @alexbr: why 50?
+#define UIO_SND_IRQ 50
+
+#define MAX_IRQ_CH 63
+int passthrough_irq_map[MAX_IRQ_CH];
+
+
+#define SERIAL_MUX_TX_CH 1
+#define SERIAL_MUX_RX_CH 2
```
Comment (Ivan-Velickovic) [inline examples/virtio-snd/snd_driver_vmm.c:63]:
```suggestion
#define SERIAL_TX_CH 1
#define SERIAL_RX_CH 2
```

Should refer to virt but it's easier to just refer to the RX and TX directions, the client doesn't care what's on the other end (if it's some other component or an actual virtualiser/mux)

File: src/util/buffer.h
Comment (alexandermbrown) [reply src/util/buffer.h:1]: removed

**Comment by Ivan-Velickovic** [comment]
Make a GitHub issue for any future work, I will start commenting what I think should go into the issue as future work.

Also, if possible any `@alexbr` should also go in there and be removed from the code.

File: examples/virtio-snd/sddf
Comment (Ivan-Velickovic) [inline examples/virtio-snd/sddf:1]: Hmm okay let's create a dep directory that contains sDDF instead of using symlinks, I'll make a PR for that now.

File: examples/virtio-snd/snd_driver_vmm.c [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+                                                  &uio_sound_fault_handler, NULL);
+    assert(success);
+
+#if defined(BOARD_qemu_arm_virt)
+    register_passthrough_irq(37, 5); // Serial
+#elif defined(BOARD_odroidc4)
+    register_passthrough_irq(48, 5); // USB controller
+    register_passthrough_irq(63, 6); // USB 1
+    register_passthrough_irq(62, 7); // USB 2
+    register_passthrough_irq(5, 8);  // Unknown
```
Comment (Ivan-Velickovic) [inline examples/virtio-snd/snd_driver_vmm.c:206]: Add to the list for future work to figure out what this IRQ is and if it is necessary.

All the USB pass-through IRQs seem necessary so that's okay.

File: tools/linux/uio_drivers/snd/main.c (hunk 5 above)
Comment (alexandermbrown) [reply tools/linux/uio_drivers/snd/main.c:15]: Done in https://github.com/au-ts/libvmm/commit/a62c2d8acf2a00684a8e4330f94793721a6744a2

File: .gitignore (hunk 4 above)
Comment (alexandermbrown) [reply .gitignore:7]: For some reason it gets put in `board/` in both `examples/virtio` and `examples/virtio-snd`

File: src/virtio/sound.c (hunk 7 above)
Comment (alexandermbrown) [reply src/virtio/sound.c]: Removed in https://github.com/au-ts/libvmm/commit/547db9f51de00395bb355f56855758b1bd4c12bb

File: examples/virtio/board/odroidc4/blk_driver_vm/dts/io.dts
Comment (Ivan-Velickovic) [inline examples/virtio/board/odroidc4/blk_driver_vm/dts/io.dts:1]: `io.dts` or `passthrough.dts`?

File: examples/virtio-snd/snd_driver_vmm.c (hunk 8 above)
Comment (alexandermbrown) [reply examples/virtio-snd/snd_driver_vmm.c]: Done in https://github.com/au-ts/libvmm/commit/772ca2a58cf66ed4e022b4fdad508b8c00930215

File: examples/virtio-snd/client_vmm.c [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+ * Copyright 2023, UNSW
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+#include <stddef.h>
+#include <stdint.h>
+#include <microkit.h>
+#include "util/atomic.h"
+#include "util/util.h"
+#include "arch/aarch64/vgic/vgic.h"
```
Comment (Ivan-Velickovic) [inline examples/virtio-snd/client_vmm.c:11]:
```suggestion
```

File: examples/virtio-snd/snd_driver_vmm.c [hunk 11]
```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright 2023, UNSW
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+#include <stddef.h>
+#include <stdint.h>
+#include <microkit.h>
+#include "util/util.h"
+#include "arch/aarch64/vgic/vgic.h"
```
Comment (Ivan-Velickovic) [inline examples/virtio-snd/snd_driver_vmm.c:10]:
```suggestion
```

File: examples/virtio/board/odroidc4/blk_driver_vm/dts/io.dts
Comment (Ivan-Velickovic) [reply examples/virtio/board/odroidc4/blk_driver_vm/dts/io.dts:1]: Nevermind, I'm wrong.

File: src/virtio/mmio.h [hunk 12]
```diff
@@ -103,18 +104,18 @@ typedef struct virtio_device_info {
...
+//     void *queue_h;
+//     /* Pointer to the sDDF config space */
+//     void *config;
+//     /* Shared data region address */
+//     uintptr_t data;
+//     /* Shared data region size */
+//     size_t data_size;
+//     /* Microkit channel to the sDDF Virtualiser */
+//     size_t ch; // @ivanv: this is microkit specific so maybe should be a callback instead or something.
+// } sddf_handler_t;
```
Comment (Ivan-Velickovic) [inline src/virtio/mmio.h:118]: This can just be deleted now right?

File: examples/virtio-snd/sddf
Comment (alexandermbrown) [reply examples/virtio-snd/sddf:1]: Done in https://github.com/au-ts/libvmm/commit/ce2e337564f79214df88e74bf64f19f8f478851f

File: examples/virtio-snd/client_vmm.c (hunk 10 above)
Comment (alexandermbrown) [reply examples/virtio-snd/client_vmm.c]: Done in https://github.com/au-ts/libvmm/commit/5c8773b6b2a131ee159d79874101cc07a498928e

File: examples/virtio-snd/pack.sh
Comment (Ivan-Velickovic) [inline examples/virtio-snd/pack.sh:1]: Is this still necessary?

File: src/virtio/sound.c [hunk 13]
```diff
@@ -0,0 +1,852 @@
+#include "sound.h"
+#include "config.h"
+#include "microkit.h"
+#include "virq.h"
+#include "virtio/mmio.h"
+#include "virtio/virtq.h"
+#include <sddf/sound/queue.h>
+
+#define DEBUG_SOUND
```
Comment (Ivan-Velickovic) [inline src/virtio/sound.c:9]: Should we comment this?

File: docs/SOUND.md
Comment (alexandermbrown) [reply docs/SOUND.md:1]: Done in https://github.com/au-ts/libvmm/commit/3cc1bc2a315e77c624db2c8305d3a15d9be319af

**Comment by Ivan-Velickovic** [comment]
Can you check `examples/virtio-snd/README.md` which I've re-done and tell me if it makes sense.

**Comment by Ivan-Velickovic** [comment]
Minor point, all good if not possible though, do you have SVGs for the diagrams instead of PNGs?

File: src/virtio/sound.c (hunk 13 above)
Comment (alexandermbrown) [reply src/virtio/sound.c]: Done in https://github.com/au-ts/libvmm/commit/2cb17ba06fc65cce1241ff3e4de821e6e85f2316

**Comment by alexandermbrown** [comment]
> Can you check `examples/virtio-snd/README.md` which I've re-done and tell me if it makes sense.

Looks good

File: .gitignore (hunk 4 above)
Comment (alexandermbrown) [reply .gitignore]: Fixed in https://github.com/au-ts/libvmm/commit/ea9d281a633201b0634762deb141e46379032724

**Comment by alexandermbrown** [comment]
> Minor point, all good if not possible though, do you have SVGs for the diagrams instead of PNGs?

Done in https://github.com/au-ts/libvmm/commit/01068d617a51dca66b01bcaefd27cfc16806487b

**Comment by alexandermbrown** [comment]
> Make a GitHub issue for any future work, I will start commenting what I think should go into the issue as future work.

Done https://github.com/au-ts/libvmm/issues/63

> Also, if possible any `@alexbr` should also go in there and be removed from the code.

Done https://github.com/au-ts/libvmm/commit/ba288af30979531897ec52fcc1cdeb9c2c79cd2b

File: src/util/atomic.h [hunk 14]
```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+// Qemu faults if you try load or store with a strict memorder on aarch64.
+#ifdef BOARD_qemu_arm_virt
```
Comment (Ivan-Velickovic) [inline src/util/atomic.h:4]: Just making a note here so I don't forget that we should return to do this. I'll chuck it in a GitHub issue soon.

File: src/virtio/console.c [hunk 15]
```diff
@@ -240,7 +249,7 @@ int virtio_console_handle_rx(struct virtio_device *dev)
         rx_queue->last_idx++;
 
         // 3. Inject IRQ to guest
-        // @ivanv: is setting interrupt status necesary?
+        // @ivanv: is setting interrupt status necessary?
```
Comment (Ivan-Velickovic) [inline src/virtio/console.c:252]: Nice :)

File: examples/virtio-snd/snd_driver_vmm.c (hunk 9 above)
Comment (alexandermbrown) [reply examples/virtio-snd/snd_driver_vmm.c:204]: done

**Review by Ivan-Velickovic — APPROVED** [review]
Fantastic work!
